### PR TITLE
Add HDMF File type

### PR DESCRIPTION
**Pull request by ajtritt**
Add a Container that has functionality for tracking foreign field and ontology information. 

This container should be used as the base type for any Container that wants to manage foreign field information or ontology information.

### CODE REVIEW

File: common/file.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+groups:
+- data_type_def: HDMFFile
+  data_type_inc: Container
+  doc: A base Container type for a logical file
+  groups:
+  - name: .ontologies
```
Comment (rly) [inline common/file.yaml:6]: I believe Lawrence said something about MatNWB not being able to handle group names that start with a dot, and he suggested underscore instead.

File: common/foreign.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+datasets:
+- data_type_def: ForeignFieldTable
+  doc: A table for identifying which objects in a file exist
+       outside the file
+  dtype:
+   - name: id
+     dtype: uint
```
Comment (rly) [inline common/foreign.yaml:7]: Minor suggestion: specify minimum precision explicitly

```suggestion
     dtype: uint32
```

Comment (rly) [reply common/foreign.yaml:7]: Also for the uints in `OntologyTable` below.

File: common/foreign.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+     doc: the namespace that data_type belongs to
+   - name: parent_object_id
+     dtype: text
+     doc: the UUID for the parent object that contains the foreign object
+   - name: object_id
+     dtype: text
+     doc: the UUID for the object that is foreign
+   - name: path
+     dtype: text
+     doc: the location of the object in the file
```
Comment (rly) [inline common/foreign.yaml:23]: An explicit format for this path should be specified, i.e., the path should use backslash as a separator before a nested group or dataset, should not start with a backslash, and should be relative to the file root. For example, if the object is named "nested_group" which is inside "parent_group" which is at the root level of the file, then the path to the object is "parent_group/nested_group".

File: common/ontology.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+     doc: the unique identifier in this table
+   - name: key
+     dtype: text
+     doc: the user key that maps to the ontology term
+   - name: ontology
+     dtype: text
+     doc: the ontology that the term comes from
+   - name: uri
+     dtype: text
+     doc: the unique resource identifier for the ontology term
```
Comment (rly) [inline common/ontology.yaml:32]: Should these tables have 
```
dims:
- num_rows
shape: 
- null
```
to restrict the tables to be 1-dimensional?

Comment (ajtritt) [reply common/ontology.yaml:32]: This is a compound type. I don’t think individual types can be more than 1D

File: common/foreign.yaml (hunk 3 above)
Comment (ajtritt) [reply common/foreign.yaml:23]: We could create a dtype for this, similar to how we have an isodatetime string

File: common/ontology.yaml (hunk 4 above)
Comment (rly) [reply common/ontology.yaml:32]: I'm confused. Initially I thought that if the shape is not specified, the dataset can have any shape, but that's wrong - it actually means the dataset is a scalar dataset. Is this meant to be a scalar dataset?

Comment (ajtritt) [reply common/ontology.yaml:32]: I misunderstood your comment. you are correct on setting dims/shape

**Comment by ajtritt** [comment]
Ontology table will be addressed by #20 

Foreign fields will not be part of the data model, since it is a feature of the backed/IO layer. See https://github.com/hdmf-dev/hdmf/pull/328 for more details.

**Pull request closed without merging**